documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+# Elemental Expression Lowering To FIR
+FIR operations are not elemental. Instead elemental expression from the parse-tree must be lowered to `fir.loop` operations
+that are looping over the array elements.
+FIR lowering of array expressions allow the implementation of the design in 
+https://github.com/flang-compiler/f18/blob/master/documentation/ArrayComposition.md.
+Lowering could implement the transformations proposed in the above mentioned document while lowering.
+However, it also may not have to assuming it will be possible to perform these transformations later on FIR.
+
+## Observations
+### Storage must be associated to array expression to lower it
+The first observation is that a `fir.loop` is more or less a regular loop and
+it is not returning an ssa value that would be the result of the array expression associated to the loop.
+Instead, a `fir.store` operation in the `fir.loop` must store array result elements into some memory reference.
+This implies that every `fir.loop` computing an array expression must be given a memory reference that can store
+the array result.
+
+
+### Scalar sub-expression must be evaluated at most once.
+In `array + cos(scalar)`, `cos(scalar)` must not be evaluated for every element of `array`.
+
+### Shape may be a compiled time constant shape or not
+In case it is not, FIR must be inserted to compute the `fir.loop` bounds.
+In case the resulting shape is empty (one dimension extent is zero),
+nothing should be evaluated in the expression unless it was evaluated to deduce the shape.
```

Comment:
It is indeed about the iteration space but there is more since we need to allocate memory to hold the result based on the shape size.
Not that clear, I will rephrase.